python/pfs/drp/stella/lsf.py

```diff
@@ -0,0 +1,688 @@
+    def interpolate(self, indices):
+        """Interpolate the kernel at provided indices
+
+        Parameters
+        ----------
+        indices : array_like, floating-point
+            Indices at which to interpolate.
+
+        Returns
+        -------
+        values : array_like, floating-point
+            Values of kernel at the provided indices.
+        """
+        return np.interp(indices, self.indices, self.values, left=0.0, right=0.0)
+
+
+class Lsf(ABC):
+    """Abstract base class for a Line-Spread Function
+
+    A line-spread function (LSF) is the one-dimensional version of a
+    point-spread function (PSF). It models the response of the spectrograph to a
+    single infinitely thin line, which may be a function of position.
+
+    This class knows nothing about wavelengths: all positions are specified in
+    pixels.
+
+    Subclasses should implement the abstract ``computeKernel`` method; other
+    methods have default implementations that rely on that.
+
+    This class includes two main styles of methods:
+
+    - methods taking a ``center`` position of type `float`. These methods are
+      intended for interacting with a numpy array.
+    - methods taking a ``point`` of type `lsst.geom.Point2D` (optionally; if not
+      provided, the ``point`` defaults to the result of ``getAveragePosition``).
+      These methods are  intended for emulating (duck-typing) an
+      `lsst.afw.detection.Psf` for when interacting with a
+      `lsst.afw.image.Image`.
+
+    Parameters
+    ----------
+    length : `int`
+        Array length.
+    interpolator : callable, optional
+        Interpolator factory for kernel. Calling this with arrays of ``x`` and
+        ``y`` should produce an interpolator that can be called with an array of
+        ``x`` values at which to interpolate.
+    """
+    def __init__(self, length, interpolator=DEFAULT_INTERPOLATOR):
+        self.length = length
+        self.interpolator = interpolator
+
+    def computeImage(self, point=None):
+        """Return an image with the LSF inserted at the nominated position
+
+        This method does the same as the ``computeArray`` method, but packages
+        the result as an ``Image``.
+
+        Parameters
+        ----------
+        point : `lsst.geom.Point2D`, optional
+            Position at which to evaluate the LSF.
+
+        Returns
+        -------
+        image : `lsst.afw.image.Image`
+            Image containing the LSF inserted.
+        """
+        if point is None:
+            point = self.getAveragePosition()
+        return ImageF(self.computeArray(point.getX())[np.newaxis, :].astype(np.float32))
+
+    def computeArray(self, center):
+        """Return an array with the LSF inserted at the nominated position
+
+        Besides the difference in return types from the ``computeKernel``
+        method (this returns an array, while ``computeKernel`` returns a
+        `Kernel1D`), this method allows positioning of the LSF at sub-pixel
```

Comment:
I don't understand.  The image based version must handle sub-pixel offsets too (as does the `Psf` class which returns the sub-image with XY0 set properly)